Tools/TestWebKitAPI/Tests/WebKitCocoa/WritingTools.mm

```diff
@@ -1198,6 +1198,92 @@ static void checkColor(WebCore::CocoaColor *color, std::optional<ColorExpectatio
     TestWebKitAPI::Util::run(&finished);
 }
 
+#if PLATFORM(MAC)
+static RetainPtr<NSAttributedString> makeTableAttributedString(NSArray<NSArray<NSString *> *> *rawTable)
+{
+    if (!rawTable.count)
+        return nil;
+
+    RetainPtr textTable = adoptNS([[NSTextTable alloc] init]);
+    [textTable setNumberOfColumns:rawTable.firstObject.count];
+
+    RetainPtr result = adoptNS([[NSMutableAttributedString alloc] init]);
+
+    for (NSUInteger row = 0; row < rawTable.count; row++) {
+        for (NSUInteger column = 0; column < rawTable[row].count; column++) {
+            RetainPtr textTableBlock = adoptNS([[NSTextTableBlock alloc] initWithTable:textTable.get() startingRow:row rowSpan:1 startingColumn:column columnSpan:1]);
+
+            RetainPtr paragraphStyle = adoptNS([[NSMutableParagraphStyle alloc] init]);
+            [paragraphStyle setTextBlocks:@[ textTableBlock.get() ]];
+
+            RetainPtr attributedString = adoptNS([[NSAttributedString alloc] initWithString:[rawTable[row][column] stringByAppendingString:@"\n"] attributes:@{
+                NSParagraphStyleAttributeName : paragraphStyle.get()
+            }]);
+
+            [result appendAttributedString:attributedString.get()];
+        }
+    }
+
+    return result;
+}
+
+TEST(WritingTools, CompositionWithTable)
+{
+    RetainPtr session = adoptNS([[WTSession alloc] initWithType:WTSessionTypeComposition textViewDelegate:nil]);
+
+    RetainPtr webView = adoptNS([[WritingToolsWKWebView alloc] initWithHTMLString:@""
+        "<body contenteditable>"
+        "   <div>Task Name, Start Date, End Date, Status</div>"
+        "   <div>Task A, 2024-01-01, 2024-01-10, In Progress</div>"
+        "   <div>Task B, 2024-01-05, 2024-01-15, Completed</div>"
+        "</body>"
+    ]);
+
+    [webView focusDocumentBodyAndSelectAll];
+
+    __block bool finished = false;
+    [[webView writingToolsDelegate] willBeginWritingToolsSession:session.get() requestContexts:^(NSArray<WTContext *> *contexts) {
+        [[webView writingToolsDelegate] writingToolsSession:session.get() didReceiveAction:WTActionCompositionRestart];
+
+        RetainPtr attributedText = makeTableAttributedString(@[
+            @[ @"Task Name", @"Start Date", @"End Date", @"Status" ],
+            @[ @"Task A", @"2024-01-01", @"2024-01-10", @"In Progress" ],
+            @[ @"Task B", @"2024-01-05", @"2024-01-15", @"Completed" ],
+        ]);
+
+        // Invoke `didReceiveText` multiple times to ensure multiple tables are *not* created.
+
+        [[webView writingToolsDelegate] compositionSession:session.get() didReceiveText:attributedText.get() replacementRange:NSMakeRange(0, 27) inContext:contexts.firstObject finished:NO];
+        TestWebKitAPI::Util::runFor(0.1_s);
```

Comment:
Didn't Megan add helper methods to avoid direct calls to `runFor`?

Reply:
Worried that this might be flaky.

Reply:
not in the case where the content is the same; I plan on addressing this, and all the other uses of runFor, in another PR which will make all the tests less flaky and more readable